### ISSUE TITLE
passenger_base_uri should remove base URI in the URI that is passed to the application

### ISSUE THREAD

**Issue by FooBarWidget** [phusion/passenger#1637]
This way it works similarly to how Nginx's `proxy_pass` works. Right now users have to deal with this themselves.

Before we implement this, we should investigate how Node.js frameworks such as Express deal with sub-URI deployments. Does Express try to strip sub-URIs too?

**Comment by OnixGH** [comment]
In Express (4.0+) you can mount applications onto one or more subpaths using a [router](http://expressjs.com/api#app.mountpath) (e.g. `app.use('/sub/path', router);`). If an app isn't already using a router this requires a number of changes.

**Comment by nschloe** [comment]
The lack of this feature prevents us from implementing a particular item for our web service. (We need to send DOIs, strings with `/` in them, to an API.)

Let me know if there's anything I can do to help out here. :)

**Comment by FooBarWidget** [comment]
@nschloe You cannot use an Express middleware or something like that?

**Comment by nschloe** [comment]
This would mean we have to set the path explicitly in the application, something we certainly would like to avoid.

**Comment by FooBarWidget** [comment]
How about installing the middleware based on an environment variable? Passenger in fact already sets the environment variable `PASSENGER_BASE_URI`. So you could do something like this:

    if (process.env.PASSENGER_BASE_URI) {
        app.use(process.env.PASSENGER_BASE_URI, router);
    }

**Comment by nschloe** [comment]
Thanks for the hint! We actually implemented your above suggestion, and it appears to work well.

What caused some confusion here is the fact that Passenger apparently does some magic when it comes to rewriting. With nginx rewrites own, the `process.env.PASSENGER_BASE_URI` isn't set, if it's off, it is set (or so it seemed).

Anyhow, all of the confusion would be resolved if Passenger implemented a rewrite. The user just tells passenger that the app is there and there and the app doesn't have to think about it.

**Comment by FooBarWidget** [comment]
I am just wondering how much of a good idea that is. What happens to paths that the application outputs? For example, suppose that the application thinks it's running on the root URL and outputs an absolute path like 'a href="/page-2"', then that link will be a broken link. But I suppose that problem exists regardless of whether we do URL rewrites or not.

**Comment by andrenarchy** [comment]
I agree that this problem is not related to URL rewriting. The absolute URL in your example would be a bug in the application (if it is mounted on a non-root location).

I really like the idea of rewrites in passenger. It would be analogous to the `proxy_pass` directive of nginx (which one would use with plain nodejs, i.e., without passenger).

**Comment by binki** [comment]
Where’s `PASSENGER_BASE_URI` in the documentation?

**Comment by CamJN** [comment]
@binki `PASSENGER_BASE_URI` isn't documented unfortunately. I'll add it to my todo list. However it is simply set to the value of [`passenger_base_uri`](https://www.phusionpassenger.com/library/config/nginx/reference/#passenger_base_uri), except where `passenger_base_uri` is empty or `"/"`. You can see this [in the code](https://github.com/phusion/passenger/blob/69878eb58cc6273397755e62a99bb7394ab66b6d/src/agent/Core/SpawningKit/Spawner.h#L828) for now.
